proposal/custom-url-resolution.md

```diff
@@ -0,0 +1,158 @@
+
+### Steps
+
+Whenever a url reference is encountered in a sass file the following steps should be executed:
+
+- A url reference is encountered in the sass file
+- From the url function we extract the parameter value
+- Rewrite the parameter value if it contains any variables ([see using variables](#using-variables))
+- Pass the parameter value to the url rewrite plugin along with the canonical url of the current sass file
+- This plugin than returns a string or calls the done callback with a string value, if it returns anything else or nothing it should throw an error ([see JavaScript API](#javascript-api))
+- The new url value that we received from the url rewrite plugin than gets used to replace the original url value
+
+_Note: if there is no urlRewrite plugin/function none of these steps should be executed._
+
+### JavaScript API
+
+At the core of remapping the url references is the JavaScript API which allows users to return a new url reference, based on the canonical url of the sass file and original url reference.
+
+The first parameter of the urlRewrite function is an object with the following values:
+
+- `file`: The canonical url of the sass file that references the url.
+- `url`: The url reference, for example with `sass-url(file://../assets/test.png)` it would be `file://../assets/test.png`.
+
+The second parameter of the urlRewrite function is an optional done callback that is used when performing asynchronous operations.
+
+This callback function takes in two parameters:
+
+- `error`: This can be null or an Error object, `null` means there is no error.
+- `url`: This is the `string` that gets used to replace the url reference. This cannot be null or anything else, otherwise sass will throw an error.
+
+Asynchronous example:
+
+```TypeScript
+let sassOptions = {
+  // Rewrite urls asynchronously
+  rewriteUrl: async (urlReference: { file: string, url: string }, done: (error: Error | null, url: string | null) => void): void => {
+    if (urlReference.file) {
+      content = await fs.readFile(path.join(path.dirname(file), url));
+    }
+
+    done(null, content ? `data:${base64(content)))}` : url);
+  }
+}
+```
+
+Synchronous example:
+
+```TypeScript
+const outDir = '/';
+
+let sassOptions = {
+  // Rewrite urls synchronously
+  rewriteUrl: (urlReference: { file: string, url: string }): string | null => {
+    if (!urlReference.file) {
+      return url;
+    }
+
+    return path.relative(path.dirname(outDir), path.join(path.dirname(file), url));
+  }
+}
+```
+
+### Url Rewrite Presets
+
+Besides writing custom plugins, there are also presets to choose from: `none`, `inline` and `relative`. The default value is `inline`.
+
+On the cli this can be configured using the flag `--url-rewrite-preset`, example: `--url-rewrite-preset=inline`
+
+In the JavaScript configuration the flag is `urlRewritePreset`, example: `{ urlRewritePreset: "none" }`
```

Comment:
Not entirely sure how you see this. How would a cli work in this case and what is the added value?

The idea I had behind using a simple string based flag is that bundlers could actually read this standardised flag and overwrite it with their own logic as most of them have memory based filesystems for caching and testing.

Reply:
The CLI would still take string names. It's just the JS API that would use callbacks.

I don't like the idea of bundlers intercepting configuration that already has a specific meaning. If they wanted to define e.g. a `parcel.url.inline` function that users could pass in instead of `sass.url.inline`, that would be just about as usable and substantially more explicit.

Reply:
Oh alright, in that case I don't really think this proposal is the place to implement that, will remove it from the proposal.

It seems like an implementation detail for bundlers and the cli in that case...

Reply:
You don't need to specify the CLI, but you should specify the built-in URL resolvers that should ship with the JS API.